proxyd/frontend_rate_limiter.go

```diff
@@ -0,0 +1,127 @@
+// timestamp and a mutex. The map is used to keep track of rate
+// limit keys, and their used limits.
+type limitedKeys struct {
+	truncTS int64
+	keys    map[string]int
+	mtx     sync.Mutex
+}
+
+func newLimitedKeys(t int64) *limitedKeys {
+	return &limitedKeys{
+		truncTS: t,
+		keys:    make(map[string]int),
+	}
+}
+
+func (l *limitedKeys) Take(key string, max int) bool {
+	l.mtx.Lock()
+	defer l.mtx.Unlock()
+	val, ok := l.keys[key]
+	if !ok {
+		l.keys[key] = 0
+		val = 0
+	}
+	l.keys[key] = val + 1
+	return val < max
+}
+
+// MemoryFrontendRateLimiter is a rate limiter that stores
+// all rate limiting information in local memory. It works
+// by storing a limitedKeys struct that references the
+// truncated timestamp at which the struct was created. If
+// the current truncated timestamp doesn't match what's
+// referenced, the limit is reset. Otherwise, values in
+// a map are incremented to represent the limit.
+type MemoryFrontendRateLimiter struct {
+	currGeneration *limitedKeys
+	dur            time.Duration
+	mtx            sync.Mutex
+}
+
+func NewMemoryFrontendRateLimit(dur time.Duration) FrontendRateLimiter {
+	return &MemoryFrontendRateLimiter{
+		dur: dur,
+	}
+}
+
+func (m *MemoryFrontendRateLimiter) Take(ctx context.Context, key string, max int) (bool, error) {
+	m.mtx.Lock()
+	// Create truncated timestamp
+	truncTS := truncateNow(m.dur)
+
+	// If there is no current rate limit map or the rate limit map reference
+	// a different timestamp, reset limits.
+	if m.currGeneration == nil || m.currGeneration.truncTS != truncTS {
+		m.currGeneration = newLimitedKeys(truncTS)
+	}
+
+	// Pull out the limiter so we can unlock before incrementing the limit.
+	limiter := m.currGeneration
+
+	m.mtx.Unlock()
+
+	return limiter.Take(key, max), nil
+}
+
+// RedisFrontendRateLimiter is a rate limiter that stores data in Redis.
+// It uses the basic rate limiter pattern described on the Redis best
+// practices website: https://redis.com/redis-best-practices/basic-rate-limiting/.
+type RedisFrontendRateLimiter struct {
+	r   *redis.Client
+	dur time.Duration
+}
+
+func NewRedisFrontendRateLimiter(r *redis.Client, dur time.Duration) FrontendRateLimiter {
+	return &RedisFrontendRateLimiter{r: r, dur: dur}
+}
+
+func (r *RedisFrontendRateLimiter) Take(ctx context.Context, key string, max int) (bool, error) {
+	var incr *redis.IntCmd
+	truncTS := truncateNow(r.dur)
+	fullKey := fmt.Sprintf("%s:%d", key, truncTS)
+	_, err := r.r.Pipelined(ctx, func(pipe redis.Pipeliner) error {
+		incr = pipe.Incr(ctx, fullKey)
+		pipe.Expire(ctx, fullKey, r.dur-time.Second)
+		return nil
+	})
+	if err != nil {
+		return false, err
+	}
+
+	return incr.Val()-1 < int64(max), nil
+}
+
+// truncateNow truncates the current timestamp
+// to the specified duration.
+func truncateNow(dur time.Duration) int64 {
+	// pointless to truncate to seconds since
+	// the UNIX timestamp is already in seconds
+	if dur == time.Second {
```

Comment:
nit: it is nice though to have branchless code.

Reply:
I was concerned with the performance of this, but turns out it doesn't really matter:

```
BenchmarkTrunc
BenchmarkTrunc/trunc
BenchmarkTrunc/trunc-8         	33962488	        35.38 ns/op
BenchmarkTrunc/now
BenchmarkTrunc/now-8           	28988336	        40.55 ns/op
PASS
``

Reply:
Ironically the `trunc` is slightly faster 🤷

Reply:
Updated to remove the branching.